python/csrc/batch_prefill.cu

```diff
@@ -310,43 +359,91 @@ std::vector<torch::Tensor> BatchPrefillWithPagedKVCachePyTorchWrapper::ForwardCu
   const LogitsPostHook logits_post_hook =
       logits_soft_cap > 0.f ? LogitsPostHook::kSoftCap : LogitsPostHook::kNone;
 
-  DISPATCH_PYTORCH_DTYPE_TO_CTYPE_FP16(q.scalar_type(), c_type, [&] {
-    return DISPATCH_logits_post_hook(logits_post_hook, LOGITS_POST_HOOK, [&] {
-      paged_kv_t<PageStorage::kIndices, c_type, int32_t> paged_kv(
-          num_kv_heads, page_size, head_dim, batch_size, kv_layout_,
-          static_cast<c_type*>(paged_kv_cache.has_value() ? paged_kv_cache->data_ptr() : nullptr),
-          static_cast<c_type*>(paged_k_cache.has_value() ? paged_k_cache->data_ptr() : nullptr),
-          static_cast<c_type*>(paged_v_cache.has_value() ? paged_v_cache->data_ptr() : nullptr),
-          static_cast<int32_t*>(paged_kv_indices.data_ptr()),
-          static_cast<int32_t*>(paged_kv_indptr.data_ptr()),
-          static_cast<int32_t*>(paged_kv_last_page_len.data_ptr()));
-      return DISPATCH_head_dim(head_dim, HEAD_DIM, [&] {
-        return DISPATCH_allow_fp16_qk_reduction(
-            allow_fp16_qk_reduction, ALLOW_FP16_QK_REDUCTION, [&] {
-              return DISPATCH_pos_encoding_mode(
-                  PosEncodingMode(pos_encoding_mode), POS_ENCODING_MODE, [&] {
-                    cudaError_t status = BatchPrefillWithPagedKVCacheWrapperDispatched<
-                        PageStorage::kIndices, HEAD_DIM, LOGITS_POST_HOOK, POS_ENCODING_MODE,
-                        ALLOW_FP16_QK_REDUCTION, MASK_MODE, c_type, c_type, int32_t>(
-                        handler_.get(), static_cast<c_type*>(q.data_ptr()),
-                        static_cast<int32_t*>(qo_indptr.data_ptr()),
-                        /*q_offset=*/nullptr, paged_kv,
-                        static_cast<uint8_t*>(custom_mask.data_ptr()),
-                        static_cast<int32_t*>(qk_indptr.data_ptr()),
-                        static_cast<c_type*>(o.data_ptr()),
-                        /*lse=*/return_lse ? static_cast<float*>(lse.data_ptr()) : nullptr,
-                        num_qo_heads, window_left, logits_soft_cap, sm_scale, rope_scale,
-                        rope_theta,
-                        /*stream=*/torch_current_stream);
-                    TORCH_CHECK(status == cudaSuccess,
-                                "BatchPrefillWithPagedKVCache failed with error code ",
-                                cudaGetErrorString(status));
-                    return true;
-                  });
-            });
+  auto q_scalar_type = q.scalar_type();
+  auto kv_scalar_type = paged_k_cache.scalar_type();
```

Comment:
paged_k_cache maybe type with std::optional<at::Tensor> instead of use paged_k_cache->scalar_type()